go/vt/vttablet/grpctmclient/client.go

```diff
@@ -1002,12 +1002,29 @@ func (client *Client) Backup(ctx context.Context, tablet *topodatapb.Tablet, req
 }
 
 // CheckThrottler is part of the tmclient.TabletManagerClient interface.
+// It always tries to use a cached client via the dialer pool as this is
+// called very frequently between tablets when the throttler is enabled in
+// a keyspace and the overhead of creating a new gRPC connection/channel
+// and dialing the other tablet every time is not practical.
 func (client *Client) CheckThrottler(ctx context.Context, tablet *topodatapb.Tablet, req *tabletmanagerdatapb.CheckThrottlerRequest) (*tabletmanagerdatapb.CheckThrottlerResponse, error) {
-	c, closer, err := client.dialer.dial(ctx, tablet)
-	if err != nil {
-		return nil, err
+	var c tabletmanagerservicepb.TabletManagerClient
+	var err error
+	if poolDialer, ok := client.dialer.(poolDialer); ok {
+		c, err = poolDialer.dialPool(ctx, tablet)
+		if err != nil {
+			return nil, err
+		}
 	}
-	defer closer.Close()
+
+	if c == nil {
```

Comment:
What are the cases where we don’t have a pooled dialer here?

Reply:
It's not that we don't have a pooled dialer, it's that dialPool returns nil. I would not expect it, from reading the function, but it's a fail-safe to prevent RPC failures in unexpected scenarios.

Reply:
@mattlord but wouldn’t there have been an error then? Can the pooled dialer return nil but also not return an error?

Reply:
It would be unexpected, yes. It's a pointer that anyone could set to nil at any time though as it's a shared object. From looking at the code it could also exist in the map but be nil as we don't check for that there.

What's so concerning about the failsafe here? I could add a comment that it should not generally happen.

Reply:
It’s that it seems very unexpected. In general on Go, you don’t nil check the return value if you have an error to indicate if it went right or wrong. If we’d do that everywhere, we would have very noisy code. 

So if the fallback can’t happen, we should imho remove it since it’s useless code. It also for example means it can never be properly tested or covered either if it can never happen. 

It’s also that the code being like this caused me to ask these questions which I think is a signal in itself too that it’s confusing 😃.

Reply:
@ajm188 what do you think? I see that this pattern came from this PR: https://github.com/vitessio/vitess/pull/8368

I can only guess that the idea was we don't want to consider a cached nil client value as an RPC level error here, but instead create a new dedicated client and gRPC connection when needed for any reason so that the RPC should always succeed even when there's something odd going on with the client cache? I wonder then if we shouldn't at least do this at the end of the dialPool function so that this nil value isn't used yet again in the future:
```
        // Get the least recently used/created cached client.
 	result := <-c 
        // If it's not nil, place it back at the end of the queue for future re-use.
 	if result != nil {
            c <- result
        }
 	return result.client, nil 
```

I'm OK doing a minor refactor here, although it's unrelated to the issue this PR is meant to address.

Reply:
@mattlord yeah, I think we can tackle this separately for sure.

Reply:
@mattlord cached_client and client are two different paradigms (you linked a PR that added the cached client whereas this pr changes client), so i'm not sure i understand your question.

as to the difference, as i understood them at the time of #8368:
1. `client` with normal (non-pooled) dialer opens a new gRPC connection for each RPC, and tears it down (along with the TCP overhead, etc.) at the end of each and every RPC
1. `client` with a pooled dialer, on first dial to a given host:port, opens N connections to that host, shoves them onto a channel, and cycles through those conns one at a time for each RPC. no connections are closed, ever, until someone calls `Close` on the TMC, at which point all connections are closed serially
1. `cached_client` is slightly different from (2), in that it dials at most 1 connection per host:port, and the most recent N connections stay alive. if an N+1th host gets dialed, we close the oldest connection by use-time, and add the new connection to the cache

hopefully that helps clarify both this thread and @shlomi-noach's above

Reply:
To summarize further: this PR is _not_ using (as I would describe it) the cached client. It is updating the plain (non-cached) client to use a pool of connections for the CheckThrottler RPC (which I agree is a good idea).

Separately, I'd like us to consider promoting the cached-client implementation to be the default over the non-cached (or, "oneshot") implementation.

Reply:
Thanks, @ajm188 ! For the cached client part, I meant that more generally — the tmclients that are cached in the buffered channel. But I agree that it’s not very precise/clear since we also have the `cachedConn` implementation in `cached_client.go`. I updated the wording in the PR to go from “cached client” to “tmclient pool” to be more clear. The linked PR https://github.com/vitessio/vitess/pull/8368 is where you added this pattern: https://github.com/vitessio/vitess/pull/8368/files#diff-c0aa163f88048bad1cd140049af3372d8f396784b4c93709fa269f82597a42bbR504 which was being discussed/debated here.